### PR TITLE
feat: use `Generations` when converting path environments to push

### PR DESCRIPTION
**Pull request by ysndr**
* Convert `PathEnvironment` into `ManagedEnvironment` using `Generations::init` (`flox push`)
* Use `Generations` to read generations metadata (`flox list`)
* Improve remote environment locking (`flox {push,pull}`)
* Add descriptive error variants for generations operations
* Add tests showing modifications of managed environments work and can be synched

### CODE REVIEW

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 1]
```diff
@@ -759,23 +765,37 @@ impl ManagedEnvironment {
...
         }
 
+        // try fast forward merge local env branch into project branch
+        self.floxmeta
+            .git
+            .push_ref(
+                ".",
+                format!(
+                    "refs/heads/{sync_branch}:refs/heads/{project_branch}",
+                    sync_branch = remote_branch_name(&self.system, &self.pointer),
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:784]: nonblocking: could add `let sync_branch =` at the top of the function

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 2]
```diff
@@ -747,7 +753,7 @@ impl ManagedEnvironment {
             .git
             .fetch_ref(
                 "origin",
-                &format!("{}:", remote_branch_name(&self.system, &self.pointer)),
+                &format!("+{0}:{0}", remote_branch_name(&self.system, &self.pointer)),
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:756]: blocking: comments need updating

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 3]
```diff
@@ -549,20 +549,26 @@ fn write_pointer_lockfile(
 
     let local_rev = match floxmeta.git.branch_hash(&local_ref) {
         Ok(local_rev) if local_rev == rev => None,
-        Ok(local_ref) => Some(local_ref),
-        // todo: think
+        Ok(local_rev) => Some(local_rev),
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:552]: blocking: update method comment
```
/// Write a pointer lockfile to the specified `lock_path`.
///
/// The current git revision of the tracked upstream repository identified by `pointer` is used for `rev`.
/// The current git revision of the local branch identified by `pointer` and `lock_path.parent()` is used for `local_rev` if that branch exists. Note that no error is returned if that branch does not exist.
```

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 4]
```diff
@@ -549,20 +549,26 @@ fn write_pointer_lockfile(
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:549]: nonblocking suggestion: at line 542, we shouldn't really be using lock_path.parent() to determine the branch_name. I think we need to pass in another parameter - maybe `local_ref: Optional<String>`. And then if that branch doesn't exist we can error. Or we could just pass the path

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 1 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:784]: yeah there are too many now

**Comment by ysndr** [comment]
for testing the push/pull aspect, until upstream floxhub supports the right tokes you can test locally with:

```
# setup a "local floxhub"
FLOX_FLOXHUB_PATH=$(mktemp -d)
mkdir -p $FLOX_FLOXHUB_PATH/$USER/floxmeta
git -C "$FLOX_FLOXHUB_PATH/$USER/floxmeta" init --bare
export __FLOX_FLOXHUB_URL="file://$FLOX_FLOXHUB_PATH"

# setup a separate place for your environments (optional, but useful to simulate different machines)
export FLOX_DATA_DIR="$(mktemp -d)"

# use flox push / pull
$ flox init --name env
$ flox install hello
$ flox push --owner $USER

# change to another machine
export FLOX_DATA_DIR="$(mktemp -d)" 
$ flox pull --remote $USER/env
```

**Comment by ysndr** [comment]
environment errors being cleaned up in #599

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 5]
```diff
@@ -544,41 +529,55 @@ impl ManagedEnvironment {
...
-        Ok(local_ref) => Some(local_ref),
-        // todo: think
-        Err(GitCommandBranchHashError::DoesNotExist) => None,
-        Err(err) => Err(ManagedEnvironmentError::GitBranchHash(err))?,
+    let local_rev = if let Some(ref local_ref) = local_ref {
+        match floxmeta.git.branch_hash(local_ref) {
+            Ok(local_rev) if local_rev == rev => None,
+            Ok(local_rev) => Some(local_rev),
+
+            Err(GitCommandBranchHashError::DoesNotExist) => None,
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/models/environment/managed_environment.rs:560]: nonblocking: now that the caller is passing in a branch name, shouldn't this be an error?

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 6]
```diff
@@ -544,41 +529,55 @@ impl ManagedEnvironment {
...
+///
+/// The lockfile stores the current git revision of the tracked upstream repository.
+/// When a local revision is specified,
+/// and the local revision is different from the remote revision,
+/// the local revision is also stored in the lockfile.
 ///
 /// When committed to a project, guarantees
 /// that the same version of the linked environment is used by all clones.
+/// When a local revision is specified,
+/// flox will **try to** check out the local revision
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/models/environment/managed_environment.rs:542]: blocking: I don't understand what you mean by check out, can you explain that?

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 7]
```diff
@@ -758,13 +763,12 @@ impl ManagedEnvironment {
...
+        let sync_branch = remote_branch_name(&self.system, &self.pointer);
+
         // Fetch the remote branch into FETCH_HEAD
         self.floxmeta
             .git
-            .fetch_ref(
-                "origin",
-                &format!("{}:", remote_branch_name(&self.system, &self.pointer)),
-            )
+            .fetch_ref("origin", &format!("+{sync_branch}:{sync_branch}"))
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/models/environment/managed_environment.rs:771]: blocking: comments in this method need updates

File: cli/tests/develop/develop/flake.lock [hunk 8]
```diff
@@ -1 +0,0 @@
-../../../../assets/templateFloxEnv/flake.lock
```
Comment (mkenigs) [inline cli/tests/develop/develop/flake.lock:1]: This cleanup was overdue 👍

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 6 above)
Comment (ysndr) [reply cli/flox-rust-sdk/src/models/environment/managed_environment.rs:542]: It's supposed to refer to `git checkout`
Not quite sure what's a better term here

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 5 above)
Comment (ysndr) [reply cli/flox-rust-sdk/src/models/environment/managed_environment.rs:560]: I guess, yes, also alignt better with the doc comment

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 6 above)
Comment (mkenigs) [reply cli/flox-rust-sdk/src/models/environment/managed_environment.rs]: Oh I just misunderstood you were talking about once the lockfile got pulled onto another machine, sorry

**Review by mkenigs — APPROVED** [review]

**Pull request merged**